File: aimux-core/tests/contract_test.rs
//! Contract tests — validate wire format consistency across languages.
//!
//! These tests load shared JSON fixtures and verify that Rust serialization
//! matches the expected wire format. The same fixtures are used by Node/Python
//! tests to ensure cross-language consistency.

use aimux_core::generate::GenerateTextOptions;
use aimux_core::message::{ModelMessage, Role};
use aimux_core::options::{TimeoutConfiguration, ToolChoice};
use aimux_core::stream_part::StreamPart;
use aimux_core::types::{FinishReasonUnified, ReasoningEffort};
use serde_json::Value;

/// A single fixture from the shared JSON file.
#[derive(serde::Deserialize)]
struct Fixture {
    name: String,
    #[serde(rename = "type")]
    ty: String,
    json: String,
    #[allow(dead_code)]
    description: String,
}

fn load_fixtures() -> Vec<Fixture> {
    let path = std::path::Path::new(env!("CARGO_MANIFEST_DIR"))
        .join("..")
        .join("contract-tests")
        .join("fixtures")
        .join("wire-format.json");
    let content = std::fs::read_to_string(&path)
        .unwrap_or_else(|e| panic!("failed to read fixtures at {path:?}: {e}"));
    serde_json::from_str(&content).expect("failed to parse fixtures JSON")
}

/// Assert that a Rust value serializes to the expected JSON.
fn assert_serialize<T: serde::Serialize>(value: &T, expected_json: &str, name: &str) {
    let actual =
        serde_json::to_string(value).unwrap_or_else(|e| panic!("{name}: serialize failed: {e}"));
    let actual_val: Value = serde_json::from_str(&actual)
        .unwrap_or_else(|e| panic!("{name}: actual is not valid JSON: {e}"));
    let expected_val: Value = serde_json::from_str(expected_json)
        .unwrap_or_else(|e| panic!("{name}: expected is not valid JSON: {e}"));
    assert_eq!(
        actual_val, expected_val,
        "fixture '{name}': serialization mismatch\n  expected: {expected_json}\n  actual:   {actual}"
    );
}

/// Assert that expected JSON deserializes to a Rust value that re-serializes
/// to the same JSON (round-trip).
#[allow(dead_code)]
fn assert_roundtrip<T: serde::Serialize + serde::de::DeserializeOwned>(
    expected_json: &str,
    name: &str,
    parse: impl Fn(&Value) -> Option<T>,
) {
    let expected_val: Value =
        serde_json::from_str(expected_json).expect("expected is not valid JSON");
    // We can't always construct the exact Rust value from the fixture,
    // so we verify round-trip: deserialize → re-serialize → compare.
    let roundtripped = match serde_json::from_str::<T>(expected_json) {
        Ok(v) => serde_json::to_string(&v).unwrap(),
        Err(_) => {
            // Some types need special handling — skip roundtrip for those
            return;
        }
    };
    let rt_val: Value = serde_json::from_str(&roundtripped).unwrap();
    assert_eq!(
        rt_val, expected_val,
        "fixture '{name}': round-trip mismatch"
    );
    // Suppress unused warning
    let _ = parse;
}

#[test]
fn tool_choice_wire_format() {
    assert_serialize(&ToolChoice::Auto, "\"auto\"", "tool_choice_auto");
    assert_serialize(&ToolChoice::None, "\"none\"", "tool_choice_none");
    assert_serialize(
        &ToolChoice::Required,
        "\"required\"",
        "tool_choice_required",
    );
    assert_serialize(
        &ToolChoice::Tool {
            tool_name: "get_weather".into(),
        },
        "{\"type\":\"tool\",\"toolName\":\"get_weather\"}",
        "tool_choice_tool",
    );
}

#[test]
fn role_wire_format() {
    assert_serialize(&Role::User, "\"user\"", "role_user");
    assert_serialize(&Role::System, "\"system\"", "role_system");
}

#[test]
fn finish_reason_unified_wire_format() {
    assert_serialize(
        &FinishReasonUnified::Stop,
        "\"stop\"",
        "finish_reason_unified_stop",
    );
}

#[test]
fn reasoning_effort_wire_format() {
    assert_serialize(&ReasoningEffort::High, "\"high\"", "reasoning_effort_high");
}

#[test]
fn generate_text_options_default_wire_format() {
    let opts = GenerateTextOptions::default();
    let json = serde_json::to_string(&opts).unwrap();
    let val: Value = serde_json::from_str(&json).unwrap();
    assert!(
        val.is_object(),
        "GenerateTextOptions should serialize to object"
    );
    // All fields should be null for default
    if let Value::Object(obj) = &val {
        for (_, v) in obj {
            assert!(v.is_null(), "default field should be null, got {v}");
        }
    }
}

/// RFC-0016 M2 true-case: `include_raw_chunks: Some(true)` round-trips on the
/// wire (companion of the `generate_text_options_include_raw_chunks_true`
/// fixture).
#[test]
fn generate_text_options_include_raw_chunks_true_wire_format() {
    let opts = GenerateTextOptions {
        include_raw_chunks: Some(true),
        ..Default::default()
    };
    let json = serde_json::to_string(&opts).unwrap();
    let val: Value = serde_json::from_str(&json).unwrap();
    assert_eq!(
        val.get("include_raw_chunks"),
        Some(&Value::Bool(true)),
        "include_raw_chunks must serialize as true, got {json}"
    );
    // Round-trip: deserialize back to the same option.
    let back: GenerateTextOptions = serde_json::from_str(&json).unwrap();
    assert_eq!(back.include_raw_chunks, Some(true));
}

/// RFC-0024: an explicit `session_id` crosses the wire as snake_case and
/// round-trips (companion of the `generate_text_options_with_session_id`
/// fixture).
#[test]
fn generate_text_options_session_id_wire_format() {
    let opts = GenerateTextOptions {
        session_id: Some("sess-1".into()),
        ..Default::default()
    };
    let json = serde_json::to_string(&opts).unwrap();
    let val: Value = serde_json::from_str(&json).unwrap();
    assert_eq!(
        val.get("session_id"),
        Some(&Value::String("sess-1".into())),
        "session_id must serialize as snake_case string, got {json}"
    );
    let back: GenerateTextOptions = serde_json::from_str(&json).unwrap();
    assert_eq!(back.session_id.as_deref(), Some("sess-1"));
}

/// Look up a fixture's expected JSON by name, failing loudly if it is gone.
///
/// This is the seam that makes Rust — the wire authority — actually consume the
/// shared fixture file instead of only hardcoding its own expectations. Without
/// it a fixture can drift away from Rust's real output and only the *other*
/// bindings go red, which points the blame at exactly the wrong side.
fn fixture_json(name: &str) -> String {
    load_fixtures()
        .into_iter()
        .find(|f| f.name == name)
        .unwrap_or_else(|| panic!("fixture '{name}' not found in wire-format.json"))
        .json
}

/// Every numeric field carries a non-null value, so each binding's declared
/// types are exercised rather than hidden behind `null`.
///
/// `top_k` is deliberately fractional. Rust is `Option<f64>` — matching the
/// upstream AI SDK, whose `topK?: number` carries no integer constraint — so a
/// binding that declares it as an integer cannot round-trip this fixture.
#[test]
fn generate_text_options_numeric_types_wire_format() {
    let opts = GenerateTextOptions {
        max_output_tokens: Some(256),
        temperature: Some(0.7),
        top_p: Some(0.95),
        top_k: Some(40.5),
        presence_penalty: Some(0.5),
        frequency_penalty: Some(-0.5),
        seed: Some(42),
        max_retries: Some(3),
        ..Default::default()
    };
    let expected = fixture_json("generate_text_options_numeric_types");
    assert_serialize(&opts, &expected, "generate_text_options_numeric_types");

    // Round-trip: the fractional top_k must survive as a fraction. An integer
    // field would truncate 40.5 to 40 and still "pass" a name-only check.
    let back: GenerateTextOptions = serde_json::from_str(&expected).unwrap();
    assert_eq!(back.top_k, Some(40.5), "top_k must round-trip as f64");
    assert_eq!(
        back.frequency_penalty,
        Some(-0.5),
        "penalties must stay signed"
    );
    assert_eq!(back.max_output_tokens, Some(256));
    assert_eq!(back.seed, Some(42));
    assert_eq!(back.max_retries, Some(3));
}

#[test]
fn stream_part_text_delta_wire_format() {
    let part = StreamPart::TextDelta {
        id: "tx1".into(),
        delta: "Hello".into(),
        provider_metadata: None,
    };
    let json = serde_json::to_string(&part).unwrap();
    let val: Value = serde_json::from_str(&json).unwrap();
    assert!(
        val.get("TextDelta").is_some(),
        "expected TextDelta variant, got {json}"
    );
}

#[test]
fn stream_part_stream_start_wire_format() {
    let part = StreamPart::StreamStart { warnings: vec![] };
    let json = serde_json::to_string(&part).unwrap();
    let val: Value = serde_json::from_str(&json).unwrap();
    assert!(
        val.get("StreamStart").is_some(),
        "expected StreamStart variant, got {json}"
    );
}

#[test]
fn stream_part_raw_wire_format() {
    let part = StreamPart::Raw {
        raw_value: serde_json::json!({ "id": "c1", "choices": [] }),
    };
    let json = serde_json::to_string(&part).unwrap();
    let val: Value = serde_json::from_str(&json).unwrap();
    assert_eq!(
        val.get("Raw").and_then(|r| r.get("raw_value")),
        Some(&serde_json::json!({ "id": "c1", "choices": [] })),
        "expected Raw variant with raw_value, got {json}"
    );
}

#[test]
fn model_message_text_wire_format() {
    let msg = ModelMessage::user("Hello");
    assert_serialize(
        &msg,
        "{\"role\":\"user\",\"content\":\"Hello\"}",
        "model_message_text",
    );
}

/// Deserialize one fixture into its Rust type, re-serialize, and compare the
/// result against the fixture's own JSON.
fn assert_fixture_matches_rust<T>(fixture: &Fixture)
where
    T: serde::Serialize + serde::de::DeserializeOwned,
{
    let parsed: T = serde_json::from_str(&fixture.json).unwrap_or_else(|e| {
        panic!(
            "fixture '{}' ({}) does not deserialize into its Rust type: {e}\n  fixture: {}",
            fixture.name, fixture.ty, fixture.json
        )
    });
    let rust_json = serde_json::to_string(&parsed).unwrap();
    let actual: Value = serde_json::from_str(&rust_json).unwrap();
    let expected: Value = serde_json::from_str(&fixture.json).unwrap();
    assert_eq!(
        actual, expected,
        "fixture '{}' does not match Rust's own serialization\n  fixture: {}\n  rust:    {}",
        fixture.name, fixture.json, rust_json
    );
}

/// Every fixture must round-trip through the Rust type it claims to describe.
///
/// Rust is the wire authority: the other bindings validate themselves against
/// this file, so a fixture describing a shape Rust never emits sends four
/// bindings chasing a fiction. The predecessor of this test only checked that
/// the fixture's JSON *parsed*, which is why the `session_id: null` drift
/// survived — `session_id` is `skip_serializing_if = "Option::is_none"`, so
/// Rust omits the key entirely.
///
/// A fixture whose `type` has no arm here fails rather than being skipped:
/// silent skipping is how a net grows holes.
#[test]
fn all_fixtures_match_rust_serialization() {
    let fixtures = load_fixtures();
    assert!(!fixtures.is_empty(), "no fixtures loaded");

    for f in &fixtures {
        match f.ty.as_str() {
            "ToolChoice" => assert_fixture_matches_rust::<ToolChoice>(f),
            "StreamPart" => assert_fixture_matches_rust::<StreamPart>(f),
            "GenerateTextOptions" => assert_fixture_matches_rust::<GenerateTextOptions>(f),
            "TimeoutConfiguration" => assert_fixture_matches_rust::<TimeoutConfiguration>(f),
            "Role" => assert_fixture_matches_rust::<Role>(f),
            "ModelMessage" => assert_fixture_matches_rust::<ModelMessage>(f),
            "FinishReasonUnified" => assert_fixture_matches_rust::<FinishReasonUnified>(f),
            "ReasoningEffort" => assert_fixture_matches_rust::<ReasoningEffort>(f),
            other => panic!(
                "fixture '{}' declares type '{other}', which has no round-trip arm in \
                 all_fixtures_match_rust_serialization — wire it up so the fixture is \
                 actually pinned to Rust",
                f.name
            ),
        }
    }
}

File: bindings/flutter/test/contract_test.dart
// Contract tests — validate the Dart binding against the shared wire-format
// fixtures in `contract-tests/fixtures/wire-format.json`, the same file the
// Rust, Go, Java, Kotlin, Python, Node and Swift contract tests read.
//
// Rust is the wire authority: it produces the JSON and the bindings consume it,
// so the direction that matters is DECODING — every fixture must decode into
// the Dart type it claims to describe. Encoding is checked as a Dart round-trip
// (decode → encode → decode) rather than by byte-comparing against the fixture,
// because Dart omits null fields that Rust emits explicitly; a byte comparison
// would fail on that alone and say nothing about type correctness.
//
// Pure serialization — no native library is loaded.

import 'dart:convert';
import 'dart:io';

import 'package:test/test.dart';
import 'package:aimux/types.dart';

/// The fixture lives at the repo root; the test's working directory depends on
/// how the suite is invoked, so try the usual spots (mirrors the Java runner).
File _fixtureFile() {
  const candidates = [
    '../../contract-tests/fixtures/wire-format.json', // cwd = bindings/flutter
    'contract-tests/fixtures/wire-format.json', // cwd = repo root
    '../contract-tests/fixtures/wire-format.json',
  ];
  for (final path in candidates) {
    final file = File(path);
    if (file.existsSync()) return file;
  }
  throw StateError('cannot find wire-format.json; tried $candidates');
}

List<Map<String, dynamic>> _loadFixtures() =>
    (jsonDecode(_fixtureFile().readAsStringSync()) as List<dynamic>)
        .cast<Map<String, dynamic>>();

void main() {
  group('shared wire-format fixtures', () {
    test('every fixture decodes into its Dart type', () {
      final fixtures = _loadFixtures();
      expect(fixtures, isNotEmpty, reason: 'no fixtures loaded');

      for (final fixture in fixtures) {
        final name = fixture['name'] as String;
        final type = fixture['type'] as String;
        final wire = jsonDecode(fixture['json'] as String);

        // Encoding a decoded fixture and decoding that again must produce the
        // same encoding. Comparing encodings rather than objects is deliberate:
        // these generated classes do not implement value equality, so `==`
        // would compare identities and pass no matter what. The generated
        // toJson() is shallow — nested models stay live objects — so each
        // encoding is normalized through jsonEncode/jsonDecode, which invokes
        // nested toJson() recursively.
        Object? norm(Object? o) => jsonDecode(jsonEncode(o));
        final Object? first;
        final Object? second;

        // A fixture type with no case here fails rather than being skipped:
        // silent skipping is how a net grows holes.
        switch (type) {
          case 'ToolChoice':
            first = norm(ToolChoice.fromJson(wire).toJson());
            second = norm(ToolChoice.fromJson(first).toJson());
          case 'StreamPart':
            first = norm(StreamPart.fromJson(wire as Map<String, dynamic>).toJson());
            second = norm(StreamPart.fromJson(first as Map<String, dynamic>).toJson());
          case 'GenerateContent':
            first = norm(GenerateContent.fromJson(wire as Map<String, dynamic>).toJson());
            second = norm(GenerateContent.fromJson(first as Map<String, dynamic>)
                    .toJson());
          case 'GenerateTextOptions':
            first = norm(GenerateTextOptions.fromJson(wire as Map<String, dynamic>)
                .toJson());
            second = norm(GenerateTextOptions.fromJson(first as Map<String, dynamic>)
                    .toJson());
          case 'TimeoutConfiguration':
            first = norm(TimeoutConfiguration.fromJson(wire as Map<String, dynamic>)
                .toJson());
            second = norm(TimeoutConfiguration.fromJson(first as Map<String, dynamic>)
                    .toJson());
          case 'ModelMessage':
            first = norm(ModelMessage.fromJson(wire as Map<String, dynamic>).toJson());
            second = norm(ModelMessage.fromJson(first as Map<String, dynamic>).toJson());
          case 'Role':
            first = norm(Role.fromJson(wire as String).toJson());
            second = norm(Role.fromJson(first as String).toJson());
          case 'FinishReasonUnified':
            first = norm(FinishReasonUnified.fromJson(wire as String).toJson());
            second = norm(FinishReasonUnified.fromJson(first as String).toJson());
          case 'ReasoningEffort':
            first = norm(ReasoningEffort.fromJson(wire as String).toJson());
            second = norm(ReasoningEffort.fromJson(first as String).toJson());
          default:
            fail(
              "fixture '$name' declares type '$type', which has no case in "
              'contract_test.dart — wire it up so the fixture is actually '
              'checked against Dart',
            );
        }

        expect(
          second,
          equals(first),
          reason:
              "fixture '$name' does not survive a Dart encode/decode round-trip",
        );
      }
    });

    test('numeric fields keep their declared precision', () {
      // `top_k` is fractional on purpose: Rust is Option<f64> (matching the
      // upstream AI SDK's `topK?: number`), so a binding that declares it an
      // integer cannot hold this value. Kotlin and Java both declared it Long
      // until this fixture was added.
      final fixture = _loadFixtures().firstWhere(
        (f) => f['name'] == 'generate_text_options_numeric_types',
        orElse: () => throw StateError(
          "fixture 'generate_text_options_numeric_types' is missing",
        ),
      );
      final opts = GenerateTextOptions.fromJson(
        jsonDecode(fixture['json'] as String) as Map<String, dynamic>,
      );

      expect(opts.topK, 40.5,
          reason: 'top_k must keep its fraction — an int would truncate to 40');
      expect(opts.frequencyPenalty, -0.5, reason: 'penalties must stay signed');
      expect(opts.temperature, 0.7);
      expect(opts.topP, 0.95);
      expect(opts.maxOutputTokens, 256);
      expect(opts.seed, 42);
      expect(opts.maxRetries, 3);
    });

    test('provider-executed transcript uses typed content parts', () {
      final fixture = _loadFixtures().firstWhere(
        (f) => f['name'] == 'model_message_provider_executed_tool_transcript',
      );
      final message = jsonDecode(fixture['json'] as String) as Map<String, dynamic>;
      final content = (message['content'] as List<dynamic>)
          .cast<Map<String, dynamic>>();
      final call = ContentPart.fromJson(content[0]) as ContentPartToolCall;
      final result = ContentPart.fromJson(content[1]) as ContentPartToolResult;

      expect(call.providerExecuted, true);
      expect(call.toolName, 'search');
      expect(result.toolName, 'search');
      expect(result.isError, false);
      expect(result.preliminary, true);
      expect(result.isDynamic, true);
    });
  });
}
